File: orchestrator/src/types/events.ts
export interface TelemetryEvent {
  event_id: string;
  timestamp: string;
  event_category: string;
  event_type: string;
  severity: string;
  actor: {
    user_id: string;
    username: string;
    department: string;
    role: string;
    source_ip: string;
    device_id?: string;
  };
  target: {
    resource_id: string;
    resource_name: string;
    resource_type: string;
    file_path?: string;
  };
  action: {
    operation: string;
    status: string;
    bytes_transferred?: number;
    session_id?: string;
  };
  ground_truth: {
    is_attack: boolean;
    scenario_id?: string | null;
    attack_phase?: string | null;
  };
  _t0_ns?: number;
}

export interface AnomalyBreakdown {
  novelty_score: number;
  temporal_anomaly: number;
  volume_max_z: number;
  pyod_percentile: number;
  peer_deviation: number;
}

export interface AnomalyResponse {
  user_id: string;
  risk_score: number;
  is_anomaly: boolean;
  severity: string;
  breakdown: AnomalyBreakdown;
  recommended_action: string;
}

export interface Alert {
  id: string;
  timestamp: string;
  severity: string;
  message: string;
  scenario_id?: string;
  acknowledged: boolean;
}

export interface UserRiskState {
  user_id: string;
  username: string;
  department: string;
  role: string;
  current_risk: number;
  peak_risk: number;
  risk_history: number[];
  anomaly_count: number;
  last_event_time: number;
  is_isolated: boolean;
  is_session_revoked: boolean;
  active_alerts: Alert[];
}

export interface WSMessage {
  type: 'log_event' | 'risk_update' | 'alert' | 'soar_action' | 'event_blocked' | 'system';
  payload: any;
  timestamp: string;
  _t0_ns?: number;
}

export interface SoarAction {
  action: 'ISOLATE_ACCOUNT' | 'REVOKE_SESSION' | 'FLAG_AUDIT' | 'STEP_UP_AUTH';
  target_user_id: string;
  analyst_note?: string;
}

export interface SoarResult {
  success: boolean;
  message: string;
  userState?: UserRiskState;
}

File: engine/app/schemas/events.py
from pydantic import BaseModel, Field, ConfigDict
from typing import Optional

class ActorInfo(BaseModel):
    user_id: str
    username: str
    department: str
    role: str
    source_ip: str
    device_id: Optional[str] = None

class TargetInfo(BaseModel):
    resource_id: str
    resource_name: str
    resource_type: str
    file_path: Optional[str] = None

class ActionInfo(BaseModel):
    operation: str
    status: str
    bytes_transferred: int = 0
    session_id: Optional[str] = None

class GroundTruth(BaseModel):
    is_attack: bool = False
    scenario_id: Optional[str] = None
    attack_phase: Optional[str] = None

class TelemetryEvent(BaseModel):
    model_config = ConfigDict(populate_by_name=True)
    event_id: str
    timestamp: str
    event_category: str
    event_type: str
    severity: str = 'INFORMATIONAL'
    actor: ActorInfo
    target: TargetInfo
    action: ActionInfo
    ground_truth: GroundTruth = GroundTruth()
    t0_ns: Optional[int] = Field(default=None, alias='_t0_ns')

class AnomalyBreakdown(BaseModel):
    novelty_score: float
    temporal_anomaly: float
    volume_max_z: float
    pyod_percentile: float
    peer_deviation: float

class AnomalyResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)
    user_id: str
    risk_score: float
    is_anomaly: bool
    severity: str
    breakdown: AnomalyBreakdown
    recommended_action: str
    t0_ns: Optional[int] = Field(default=None, alias='_t0_ns')
